src/NSubstitute/Core/CallRouter.cs

```diff
@@ -65,59 +55,50 @@ public IEnumerable<ICall> ReceivedCalls()
 
+            var isQuerying = _threadContext.IsQuerying;
             var pendingRaisingEventArgs = _threadContext.UsePendingRaisingEventArgumentsFactory();
+            var queuedNextRouteFactory = _threadContext.UseNextRoute(this);
+
+            IRoute routeToUse = ResolveCurrentRoute(call, isQuerying, pendingRaisingEventArgs, queuedNextRouteFactory);
 
-            IRoute routeToUseForThisCall;
-            if (_threadContext.IsQuerying)
+            return routeToUse.Handle(call);
+        }
+
+        private IRoute ResolveCurrentRoute(ICall call, bool isQuerying, Func<ICall, object[]> pendingRaisingEventArgs, Func<ISubstituteState, IRoute> queuedNextRouteFactory)
+        {
+            if (isQuerying)
             {
-                routeToUseForThisCall = GetQueryRoute();
+                return _routeFactory.CallQuery(_substituteState);
             }
-            else if (pendingRaisingEventArgs != null)
+
+            if (pendingRaisingEventArgs != null)
             {
-                routeToUseForThisCall = GetRaiseEventRoute(pendingRaisingEventArgs);
+                return _routeFactory.RaiseEvent(_substituteState, pendingRaisingEventArgs);
             }
-            else if (IsSpecifyingACall(call, _currentRoute))
+
+            if (queuedNextRouteFactory != null)
             {
-                routeToUseForThisCall = GetRecordCallSpecRoute();
+                return queuedNextRouteFactory.Invoke(_substituteState);
             }
-            else
+
+            if (IsSpecifyingACall(call))
             {
-                routeToUseForThisCall = _currentRoute;
+                return _routeFactory.RecordCallSpecification(_substituteState);
             }
 
-            UseDefaultRouteForNextCall();
-            return routeToUseForThisCall.Handle(call);
-        }
-
-        private IRoute GetQueryRoute()
-        {
-            return _routeFactory.CallQuery(_substituteState);
-        }
-
-        private IRoute GetRaiseEventRoute(Func<ICall, object[]> argumentsFactory)
-        {
-            return _routeFactory.RaiseEvent(_substituteState, argumentsFactory);
-        }
-
-        private static bool IsSpecifyingACall(ICall call, IRoute currentRoute)
-        {
-            var args = call.GetOriginalArguments() ?? EmptyArgs;
-            var argSpecs = call.GetArgumentSpecifications() ?? EmptyArgSpecs;
-            return currentRoute.IsRecordReplayRoute && args.Any() && argSpecs.Any();
-        }
-
-        private IRoute GetRecordCallSpecRoute()
-        {
-            return _routeFactory.RecordCallSpecification(_substituteState);
+            return _routeFactory.RecordReplay(_substituteState);
         }
 
-        private void UseDefaultRouteForNextCall()
+        private static bool IsSpecifyingACall(ICall call)
         {
-            SetRoute(x => _routeFactory.RecordReplay(x));
+            return call.GetOriginalArguments().Length != 0 && call.GetArgumentSpecifications().Count != 0;
```

Comment:
The previous implementation also checked `currentRoute.IsRecordReplayRoute`. Is this one equivalent?

Reply:
I analyzed the flow and found that `route.IsRecordReplay` check is redundant now.

Before the code was like this:

```c#
            IRoute routeToUseForThisCall;
            if (_threadContext.IsQuerying)
            {
                routeToUseForThisCall = GetQueryRoute();
            }
            else if (pendingRaisingEventArgs != null)
            {
                routeToUseForThisCall = GetRaiseEventRoute(pendingRaisingEventArgs);
            }
            else if (IsSpecifyingACall(call, _currentRoute))
            {
                routeToUseForThisCall = GetRecordCallSpecRoute();
            }
            else
            {
                routeToUseForThisCall = _currentRoute;
            }
```

There we didn't know what actually was stored in the `_currentRouter` variable. There might be the explicitly configured route, or the default `RecordReplay` one. That's why we had a special check - if user configured a custom route, we wanted to use it instead.

In my PR we consume explicitly configured next route from the thread local context. If variable is not `null` - we have a custom route, so we use it instead of default behavior. This way we don't need to check for a flag.

There is a small change in the behavior, but I don't think it matters. Previously if you _explicitly_ set `RecordReplay` route as the next one, it could still be transformed to the `RecordCallSpecRoute` during the dispatch. Now if you explicitly set `RecordReplay` route as the next one, it will be used for sure. In fact, we never set the `RecordReplay` route as the next (as it's the default one), but I like how it behaves now - if you for some reason want next route to be `RecordReplay` , then library should blindly do it for you. As usual, common sense is required 😅